design-docs/java-annotation-processing.md

```diff
@@ -44,74 +48,125 @@ public class CompileOptions extends AbstractOptions {
 
-Add a `processorpath` (similar to the existing `sourcepath`), `processors`, `processorArgs`, and `proc`:
+Add `annotationProcessing`, `processorpath` (similar to the existing `sourcepath`), `processors`, and `processorArgs` to `CompileOptions`:
 
+ * `annotationProcessing` as a `boolean`: when set to `false`, `-proc:none` is passed to the Java compiler; defaults to `true`.
  * `processorpath` as a `FileCollection`: when non-empty, it is passed as a `-processorpath` argument to the Java compiler.
  * `processors` as a `List<String>` (or `Set<String>` or `Collection<String>`?); when non-empty, it is passed as a `-processor` argument to the Java compiler (values joined with a `,`).
  * `processorArgs` as a `Map<String, ?>` (or `Map<String, String>`?); each entry is passed as a `-Akey=value` argument to the Java compiler.
- * `proc` as an enum with values `none` and `only`, defaults to `null`; when non-`null`, it is passed a `-proc` argument to the Java compiler (`-proc:none` or `-proc:only`).
 
+Note: `ApiGroovyCompiler` must be updated to take into account the `annotationProcessing` property in addition to a `-proc:none` explicit item in `compilerArgs`.
 
 #### Backwards compatibility
 
  * The new arguments must be added before the `compilerArgs` so that user-specified arguments in `compilerArgs` override the ones added by those new properties (similar to how `compilerArgs` can override `sourcepath` added in Gradle 2.4, or any other compiler argument
  * An empty `processorpath` must not result in a `-processorpath` argument so that the default behavior still is to lookup annotation processors in the classpath.
+ * `annotationProcessing` must default to `true`.
 
 #### Test cases
 
  * The default behavior doesn't change (no change in existing tests, e.g. in `JavaCompilerArgumentsBuilderTest` and `CommandLineJavaCompilerArgumentsGeneratorTest`)
  * a non-empty `processorpath` adds a `-processorpath` argument with `processorpath.asPath` as value
  * a non-empty `processors` list adds a `-processor` argument with the `processors` joined with a `,` as value
  * each entry in `processorArgs` is added as `-Akey=value` arguments
- * `proc` with a value of `none` is added as a `-proc:none` argument
- * `proc` with a value of `only` is added as a `-proc:only` argument
+ * `annotationProcessing` set to `false` is added as a `-proc:none` argument
  * Add integration test with an annotation processor in the `compileOnly` configuration and an empty `compileJava.processorpath`, and verify that it ran (checks non-regression)
  * Add integration test with an annotation processor in the `compileJava.options.processorpath` and another in the `compileOnly` configuration, and verify that the former ran but the latter didn't
 
-### Story - Add processor path to `SourceSet` (TBD if this should live in gradle core)
+### Story - Allow specification of generated sources destination directory for compile tasks
```

Comment:
AFAICT, this would work the same as for generated classes and resources in the `destinationDir`.

If you ask me, you'd better make small compile tasks (compiling fewer files) to benefit from incremental build (whether the task is up-to-date) than relying on incremental compilation (tracking dependencies between classes is a PITA, and if you add annotation processing to the equation is can easily become a nightmare).
Believe me, if compile time ever becomes an issue for me, I'd rather switch to Bazel than enable incremental compilation.
YMMV.